mat64/list_test.go

```diff
@@ -0,0 +1,706 @@
+func testTwoInput(c *check.C,
+	receiver Matrix,
+	method func(receiver, a, b Matrix),
+	denseComparison func(receiver, a, b *Dense),
+	legalTypes func(a, b Matrix) bool,
+	legalSize func(ar, ac, br, bc int) bool,
+) {
+	strideVec := &Vector{
+		mat: blas64.Vector{
+			Inc: 10,
+		},
+	}
+	// Loop over all of the matrix types.
+	matrices := []Matrix{
+		&Dense{},
+		&SymDense{},
+		NewTriDense(0, true, nil),
+		NewTriDense(0, false, nil),
+		NewVector(0, nil),
+		strideVec,
+		&basicMatrix{},
+		&basicVectorer{},
+		&basicSymmetric{},
+		&basicTriangular{},
+	}
+
+	for _, aMat := range matrices {
+		for _, bMat := range matrices {
+			// Loop over all of the size combinations (bigger, smaller, etc.)
+			for _, test := range []struct {
+				ar, ac, br, bc int
+			}{
+
+				{1, 1, 1, 1},
+				{6, 6, 6, 6},
+				{7, 7, 7, 7},
+
+				{1, 1, 1, 5},
+				{1, 1, 5, 1},
+				{1, 5, 1, 1},
+				{5, 1, 1, 1},
+				{6, 6, 6, 11},
+				{6, 6, 11, 6},
+				{6, 11, 6, 6},
+				{11, 6, 6, 6},
+				{11, 11, 11, 6},
+				{11, 11, 6, 11},
+				{11, 6, 11, 11},
+				{6, 11, 11, 11},
+
+				{1, 1, 5, 5},
+				{1, 5, 1, 5},
+				{1, 5, 5, 1},
+				{5, 1, 1, 5},
+				{5, 1, 5, 1},
+				{5, 5, 1, 1},
+				{6, 6, 11, 11},
+				{6, 11, 6, 11},
+				{6, 11, 11, 6},
+				{11, 6, 6, 11},
+				{11, 6, 11, 6},
+				{11, 11, 6, 6},
+
+				{1, 1, 17, 11},
+				{1, 1, 11, 17},
+				{1, 11, 1, 17},
+				{1, 17, 1, 11},
+				{1, 11, 17, 1},
+				{1, 17, 11, 1},
+				{11, 1, 1, 17},
+				{17, 1, 1, 11},
+				{11, 1, 17, 1},
+				{17, 1, 11, 1},
+				{11, 17, 1, 1},
+				{17, 11, 1, 1},
+
+				{6, 6, 1, 11},
+				{6, 6, 11, 1},
+				{6, 11, 6, 1},
+				{6, 1, 6, 11},
+				{6, 11, 1, 6},
+				{6, 1, 11, 6},
+				{11, 6, 6, 1},
+				{1, 6, 6, 11},
+				{11, 6, 1, 6},
+				{1, 6, 11, 6},
+				{11, 1, 6, 6},
+				{1, 11, 6, 6},
+
+				{6, 6, 17, 1},
+				{6, 6, 1, 17},
+				{6, 1, 6, 17},
+				{6, 17, 6, 1},
+				{6, 1, 17, 6},
+				{6, 17, 1, 6},
+				{1, 6, 6, 17},
+				{17, 6, 6, 1},
+				{1, 6, 17, 6},
+				{17, 6, 1, 6},
+				{1, 17, 6, 6},
+				{17, 1, 6, 6},
+
+				{6, 6, 17, 11},
+				{6, 6, 11, 17},
+				{6, 11, 6, 17},
+				{6, 17, 6, 11},
+				{6, 11, 17, 6},
+				{6, 17, 11, 6},
+				{11, 6, 6, 17},
+				{17, 6, 6, 11},
+				{11, 6, 17, 6},
+				{17, 6, 11, 6},
+				{11, 17, 6, 6},
+				{17, 11, 6, 6},
+			} {
+				// Skip the test if the arguments are compile-time provably incorrect
+				// This is either because the matrix types are wrong at compile time
+				// or because the sizes are impossible
+				ok := legalTypes(aMat, bMat)
+				if !ok {
+					continue
+				}
+				if ok := dimsLegal(aMat, test.ar, test.ac); !ok {
+					continue
+				}
+				if ok := dimsLegal(bMat, test.br, test.bc); !ok {
+					continue
+				}
+				a := makeRandOf(aMat, test.ar, test.ac)
+				{
+					ar, ac := a.Dims()
+					if ar != test.ar || ac != test.ac {
+						panic("makeRandOf returns wrong size" + fmt.Sprintf("%T\n", aMat))
```

Comment:
You can `panic(fmt.Sprintf("makeRandOf for %T returns wrong size: %dx%d != %dx%d", aMat, ar, ac, test.ar, test.ac))`. This is more informative and cleaner than appending a string to the output of fmt.Sprintf.